source/code/plugins/out_oms_diag.rb

```diff
@@ -0,0 +1,112 @@
+module Fluent
+
+  class OutputOMSDiag < BufferedOutput
+
+    Plugin.register_output('out_oms_diag', self)
+
+    def initialize
+      super
+      require 'net/http'
+      require 'net/https'
+      require 'uri'
+      require_relative 'omslog'
+      require_relative 'oms_configuration'
+      require_relative 'oms_common'
+      require_relative 'oms_diag'
+    end
+
+    config_param :omsadmin_conf_path, :string, :default => '/etc/opt/microsoft/omsagent/conf/omsadmin.conf'
+    config_param :cert_path, :string, :default => '/etc/opt/microsoft/omsagent/certs/oms.crt'
+    config_param :key_path, :string, :default => '/etc/opt/microsoft/omsagent/certs/oms.key'
+    config_param :proxy_conf_path, :string, :default => '/etc/opt/microsoft/omsagent/proxy.conf'
+    # TODO: To make compress default to true once supported
+    config_param :compress, :bool, :default => false
+
+    def configure(conf)
+      super
+    end
+
+    def start
+      super
+      @proxy_config = OMS::Configuration.get_proxy_config(@proxy_conf_path)
+    end
+
+    def shutdown
+      super
+    end
+
+    def handle_record(ipname, record)
+      @log.trace "Handling diag records for ipname : #{ipname}"
+      req = OMS::Common.create_ods_request(OMS::Configuration.diagnostic_endpoint.path, record, @compress)
+      unless req.nil?
+        http = OMS::Common.create_ods_http(OMS::Configuration.diagnostic_endpoint, @proxy_config)
+        start = Time.now
+
+        # This method will raise on failure alerting the engine to retry sending this data
+        OMS::Common.start_request(req, http)
+
+        ends = Time.now
+        time = ends - start
+        count = record[OMS::Diag::RECORD_DATAITEMS].size
+        @log.debug "Success sending diag #{ipname} x #{count} in #{time.round(2)}s"
+        return true
+      end
+    rescue OMS::RetryRequestException => e
+      @log.info "Encountered retryable exception. Will retry sending diag data later."
+      @log.debug "Error with diag:'#{e}'"
+      # Re-raise the exception to inform the fluentd engine we want to retry sending this chunk of data later.
+      raise e.message
+    rescue => e
+      # We encountered something unexpected. We drop the data because
+      # if bad data caused the exception, the engine will continuously
+      # try and fail to resend it. (Infinite failure loop)
+      OMS::Log.error_once("Unexpecting exception, dropping diag data. Error:'#{e}'")
+    end
+
+    # This method is called when an event reaches to Fluentd.
+    # Convert the event to a raw string.
+    def format(tag, time, record)
+      if record != {}
+        @log.trace "Buffering diag #{tag}"
+        retval = record.to_msgpack
+        return retval
+      else
+        return ""
+      end
+    end
+
+    def write(chunk)
```

Comment:
Well the aggregation is based on IPName alone as the diagnostics channel has that alone is mandatory (as per the postdaitaitems endpoint implementation). So all dataitems pertaining to a ipname are aggregating and handled together.
If there are spurious dataitems (dataitems constitute the chunk) which have missing logmessage or time in the hash then they are dropped and remaining are aggregated. the unit test file out_oms_diag_plugintest.rb will give you detailed clarity on this.